### ISSUE TITLE
Add Ref Forwarding to core components

### ISSUE THREAD

**Issue by dgiraldo313** [priceline/design-system#479]
It looks like we are currently not forwarding refs on our core components. This can be problematic for accessibility purposes, since it won't allow us to do things like restore focus to the element by using refs.

React introduced `forwardRef` API in v16.3, so this should make it pretty easy to update our core components. 

https://reactjs.org/blog/2018/03/29/react-v-16-3.html#forwardref-api

```js
const FancyButton = React.forwardRef((props, ref) => (
  <button ref={ref} className="FancyButton">
    {props.children}
  </button>
));

// You can now get a ref directly to the DOM button:
const ref = React.createRef();
<FancyButton ref={ref}>Click me!</FancyButton>;
```

The thing we should probably discuss as a group is which core components should receive this update and why?

I think for the sake of accessibility, the following components should be forwarding refs:
- Button
- Link
- Input
- Radio
- Select
- Slider
- Tooltip

Any others we can think of?

**Comment by jxnblk** [comment]
FWIW styled-components forwards refs by default, so you shouldn't need to change any of those components, see https://www.styled-components.com/docs/advanced#refs

**Comment by dgiraldo313** [comment]
@jxnblk That is a good point. If they are all wrapped with styled components, then we wouldn't need to forward refs. However, we would probably need to update styled components to v4. It looks like we're currently using V3, and this version uses `innerRef` prop instead of `ref`.

https://www.styled-components.com/docs/api#deprecated-innerref-prop

**Comment by BeniCheni** [comment]
@dgiraldo313 , @jxnblk , looks like `styled-components` v4 is a candidate for `next` branch for DSv3 to strategically resolve the `innerRef` support deprecation from `styled-components`, and doubt of forward ref general support.

Think we could create a prioritized issue and take it up the WG for a focused topic.

**Comment by sdalonzo** [comment]
in the meantime, this would also require marking React 16.3 as the peerDep floor (not saying we shouldn't, because we should - just making the point)

**Comment by BeniCheni** [comment]
@sdalonzo. @dgiraldo313, I've changed this issue' label to "help wanted" & "good first issue" to add React 16.3.* as a `peerDependencies` in `package.json`.

We can update the issue to capture further if you have additional thoughts to move forward with this issue.

**Comment by dansternfeld1** [comment]
hey @bertya -- we need to confirm what the actual current React minimum version is based on what features we're using, and we suspect the Modal requires 16.3

**Comment by bertya** [comment]
@dansternfeld1 modal is on 16.5.x, popover is on 16.3.x, all other packages we have is on peer 16.8.x

**Comment by craigpalermo** [comment]
Hey @dgiraldo313 @sdalonzo, what do you think about this pattern for passing a ref to a DS component that exports a styled component? Replacing the `dsRef` prop with `ref` doesn't seem to work because React handles the `ref` prop differently than other props.
<img width="1241" alt="Screen Shot 2020-04-21 at 1 47 36 PM" src="https://user-images.githubusercontent.com/3260642/79896922-faab6d00-83d6-11ea-9436-8dba6827f6e9.png">

I think this is necessary in part because the immediate child of `<Button>` is a `ForwardRef`, and so it expects the `ref` prop. This should have the advantage of behaving the same for both SC3 and 4.
<img width="1632" alt="Screen Shot 2020-04-21 at 1 50 33 PM" src="https://user-images.githubusercontent.com/3260642/79897178-5118ab80-83d7-11ea-92f2-32ecabd8a8ef.png">
